src/extra.c

```diff
@@ -117,6 +145,51 @@ SEXP ps__throw_error(void) {
   return out;
 }
 
+// Need to jump out of the `R_UnwindProtect()` context
+#include <setjmp.h>
+
+static
+SEXP check_for_user_interrupt_callback(void *_payload) {
+  R_CheckUserInterrupt();
+  return R_NilValue;
+}
+
+static
+void check_for_user_interrupt_cleanup(void *payload,
+                                      Rboolean jump) {
+  if (jump) {
+    jmp_buf *env = (jmp_buf *) payload;
+    longjmp(*env, 1);
+  }
+}
+
+SEXP r_check_for_user_interrupt(SEXP cont) {
+  jmp_buf env;
+
+  if (setjmp(env)) {
+    return cont;
+  }
+
+  R_UnwindProtect(&check_for_user_interrupt_callback,
+                  NULL,
+                  &check_for_user_interrupt_cleanup,
+                  &env,
+                  cont);
+
+  return NULL;
+}
+
+r_no_return
+void r_unwind(SEXP x) {
+  if (inherits(x, "error")) {
+    SEXP call = PROTECT(lang2(install("stop"), x));
+    eval(call, R_BaseEnv);
+  } else {
+    R_ContinueUnwind(x);
+  }
+  PS__STOP_UNREACHABLE();
+}
+
```

Comment:
This is the mechanism I mentioned on Slack that converts longjumps to continuation token. `r_unwind()` is used to resume an error path using R longjumps. It takes continuation tokens captured from jumpy functions or error objects that we created.